File: Java/2023/main/year2023/Day05/LocationCascadingMapper.java
package year2023.Day05;

import java.io.*;
import java.util.*;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import JavaDataModel.*;

@AoCSolution()
public class LocationCascadingMapper implements ISolution {
    List<Long> Seeds;

    List<List<TenaryTuple<Long, Long, Long>>> MapBlocks;

    long SearchMap(long source, int mId) {
        if (mId >= MapBlocks.size()) {
            return source;
        }
        for (var rangeInfo : MapBlocks.get(mId)) {
            long from = rangeInfo.getE1(), to = rangeInfo.getE2(), range = rangeInfo.getE3();
            if (source >= from && source < from + range) {
                return SearchMap(source - from + to, mId + 1);
            }
        }
        return SearchMap(source, mId + 1);
    }

    void readFile(BufferedReader input) throws IOException {
        MapBlocks = new ArrayList<>();
        Stream.of(input.lines().map(l -> l + "\n").reduce("", String::concat).split("\\n\\n")).forEach(block -> {
            if (block.contains("seeds")) {
                Seeds = Stream.of(block.split(":")[1].trim().split("\\s+")).map(Long::parseLong).toList();
            } else {
                MapBlocks.add(Stream.of(block.split("\\n"))
                        .filter(line -> !line.contains("map"))
                        .map(line -> Stream.of(line.split("\\s+")).map(Long::parseLong).toArray(Long[]::new))
                        .map(arr -> new TenaryTuple<>(arr[1], arr[0], arr[2])).toList());
            }
        });
    }

    public void Solution1(BufferedReader input) throws IOException {
        readFile(input);
        System.out.println("Solution 1: " + Seeds.stream()
                .mapToLong(token -> SearchMap(token, 0)).reduce(Long.MAX_VALUE, Long::min));
    }

    List<Pair<Long, Long>> SearchInterval(int mId, List<Pair<Long, Long>> itvList) {
        if (mId >= MapBlocks.size()) {
            return itvList;
        }

        List<Pair<Long, Long>> nextItvList = itvList.stream().flatMap(itv -> {
            List<Pair<Long, Long>> matchedList = new ArrayList<>(), unmatchedList = new ArrayList<>(List.of(itv));
            MapBlocks.get(mId).stream().takeWhile(__ -> !unmatchedList.isEmpty()).forEach(rangeInfo -> {
                long from = rangeInfo.getE1(), to = rangeInfo.getE2(), fromEnd = from + rangeInfo.getE3();
                List<Pair<Long, Long>> processList = new ArrayList<>(unmatchedList);
                unmatchedList.clear();
                for (Pair<Long, Long> curItv : processList) {
                    long itvStart = curItv.getKey(), itvEnd = curItv.getValue();
                    Pair<Long, Long> mapped = new Pair<>(Long.max(itvStart, from), Long.min(itvEnd, fromEnd));
                    if (mapped.getKey() >= mapped.getValue()) {
                        unmatchedList.add(curItv);
                        continue;
                    }
                    matchedList.add(new Pair<>(mapped.getKey() - from + to, mapped.getValue() - from + to));
                    // *----|------------|----*
                    // *----|------*-----|----
                    // -----|------*-----|----*
                    if (itvStart < from) {
                        unmatchedList.add(new Pair<>(itvStart, from));
                    }
                    if (itvEnd > fromEnd) {
                        unmatchedList.add(new Pair<>(fromEnd, itvEnd));
                    }
                }
            });
            matchedList.addAll(unmatchedList);
            return matchedList.stream();
        }).toList();
        return SearchInterval(mId + 1, nextItvList);
    }

    public void Solution2(BufferedReader input) throws IOException {
        System.out.println("Solution 2: " + IntStream.range(0, Seeds.size()).filter(i -> i % 2 == 0).mapToLong(i -> {
            return SearchInterval(0, List.of(new Pair<>(Seeds.get(i), Seeds.get(i) + Seeds.get(i + 1))))
                    .stream().map(Pair::getKey).reduce(Long.MAX_VALUE, Long::min);
        }).reduce(Long.MAX_VALUE, Long::min));
    }

    public static void main(String[] args) throws IOException {
        LocationCascadingMapper Day05 = new LocationCascadingMapper();
        Day05.Solution1(new BufferedReader(new FileReader("Day05/input.txt")));
        Day05.Solution2(new BufferedReader(new FileReader("Day05/input.txt")));
    }
}

File: Java/2023/main/year2023/Day05/LocationCascadingMapper.kt
package year2023.Day05

import java.io.*

import JavaDataModel.ISolution
import JavaDataModel.AoCSolution

typealias Interval = Pair<Long, Long>

@AoCSolution()
class LocationCascadingMapper : ISolution {
    lateinit var Seeds: List<Long>

    val MapBlocks: MutableList<List<Triple<Long, Long, Long>>> = mutableListOf()

    // there will be only one level stack for this search, no need further process of returned value, apply tail recursion optimize
    tailrec fun SearchMap(source: Long, level: Int): Long {
        if (level >= MapBlocks.size) {
            return source
        }
        val nextSourceOrCurrent = MapBlocks[level].firstNotNullOfOrNull { (from, to, range) ->
            if (source in from.until(from + range)) source - from + to else null
        } ?: source
        return SearchMap(nextSourceOrCurrent, level + 1)
    }

    fun readFile(input: BufferedReader) {
        input.lineSequence().joinToString("\n").splitToSequence("\n\n").forEach { block ->
            when {
                block.startsWith("seeds:") -> Seeds = block.substringAfter(":").trim()
                    .splitToSequence(Regex("\\s+")).map { it.toLong() }.toList()

                else ->
                    MapBlocks += block.lineSequence().filterNot { it.endsWith("map:") }
                        .map {
                            it.splitToSequence(Regex("\\s+")).map(String::toLong).toList()
                                .let { (to, from, range) -> Triple(from, to, range) }
                        }.toList()
            }
        }
    }

    override fun Solution1(input: BufferedReader) {
        readFile(input)
        // minBy returns T, minOf returns R
        println("Solution 1: ${Seeds.minOf { token -> SearchMap(token, 0) }}")
    }

    tailrec fun SearchInterval(level: Int, itvList: List<Pair<Long, Long>>): List<Pair<Long, Long>> {
        if (level >= MapBlocks.size) {
            return itvList
        }
        val nextItvList: List<Pair<Long, Long>> = itvList.asSequence().flatMap { itv ->
            val matchedList: MutableList<Pair<Long, Long>> = ArrayList()
            var unmatchedList: MutableList<Pair<Long, Long>> = mutableListOf(itv)
            for ((from, to, range) in MapBlocks[level]) {
                if (unmatchedList.isEmpty()) {
                    break
                }
                val fromEnd = from + range
                unmatchedList.fold(mutableListOf<Interval>() to mutableListOf<Interval>()) { (mAcc, unMAcc), (itvStart, itvEnd) ->
                    // coerceAtLeast set a lower bound for both, is Max. Ehh hard to understand
                    // val mapped = Pair(itvStart.coerceAtLeast(from), itvEnd.coerceAtMost(fromEnd))
                    val mapped = maxOf(itvStart, from) to minOf(itvEnd, fromEnd)
                    if (mapped.first >= mapped.second) {
                        unMAcc += itvStart to itvEnd
                    } else {
                        mAcc += Pair(mapped.first - from + to, mapped.second - from + to)
                        // *----|------------|----*
                        // *----|------*-----|----
                        // -----|------*-----|----*
                        if (itvStart < from) {
                            unMAcc += itvStart to from
                        }
                        if (itvEnd > fromEnd) {
                            unMAcc += fromEnd to itvEnd
                        }
                    }
                    mAcc to unMAcc
                }.apply {
                    unmatchedList = second
                    matchedList += first
                }
            }
            (matchedList + unmatchedList).asSequence()
        }.toList()
        return SearchInterval(level + 1, nextItvList)
    }

    override fun Solution2(input: BufferedReader) {
        // val f: (Long, Long) -> Long = Long::coerceAtMost // Eh, IDEA doesn't provide auto-completion on this
        println("Solution 2: " + Seeds.asSequence().chunked(2).flatMap { (begin, range) ->
            SearchInterval(0, listOf(begin to begin + range)).asSequence().map { it.first }
        }.min())
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            LocationCascadingMapper().apply {
                BufferedReader(FileReader("Day05/input.txt")).use(::Solution1)
                BufferedReader(FileReader("Day05/input.txt")).use(::Solution2)
            }
        }
    }
}
